### PR TITLE
Feature: Tasks (core)

### PR DESCRIPTION
**Pull request by abist**
Fixes https://github.com/microsoft/azuredatastudio/issues/11725

This enables the tasks system in core without the database projects build tasks. 

Had to add a problemMatcher to sql-database-projects because the tasks system needs atleast one extension with a problem matcher to light up. If we want to move this problem matcher elsewhere, such as mssql, let me know.

### CODE REVIEW

**Comment by abist** [comment]
The build is failing because of some code scanning failures that were fixed 5 days ago. I want to avoid merging main to this branch right now to avoid those changes from the actual feature changes. Once this is approved, I'll merge main to get the build fixed.

**Comment by abist** [comment]
@kburtram @alanrenmsft @udeeshagautam @Charles-Gagnon @chlafreniere gentle ping

File: src/vs/workbench/services/configuration/common/configuration.ts [hunk 1]
```diff
@@ -5,7 +5,9 @@
 
 import { ConfigurationScope } from 'vs/platform/configuration/common/configurationRegistry';
 
+// {{SQL CARBON EDIT}}
 export const FOLDER_CONFIG_FOLDER_NAME = '.azuredatastudio';
+export const TASKS_FOLDER_CONFIG_FOLDER_NAME = '.vscode';
```
Comment (alanrenmsft) [inline src/vs/workbench/services/configuration/common/configuration.ts:10]: why do we need .vscode folder?

Comment (abist) [reply src/vs/workbench/services/configuration/common/configuration.ts:10]: so users can seamlessly sync with their vscode tasks with ADS

File: src/vs/workbench/api/common/extHost.api.impl.ts [hunk 2]
```diff
@@ -755,7 +756,10 @@ export function createApiFactoryAndRegisterActors(accessor: ServicesAccessor): I
 				return extHostDocumentContentProviders.registerTextDocumentContentProvider(scheme, provider);
 			},
 			registerTaskProvider: (type: string, provider: vscode.TaskProvider) => {
-				throw new Error('Tasks api is not allowed in Azure Data Studio'); // {{SQL CARBON EDIT}} disable task
+				extHostApiDeprecation.report('window.registerTaskProvider', extension,
```
Comment (udeeshagautam) [inline src/vs/workbench/api/common/extHost.api.impl.ts:759]: Just for my understanding if we never had this what are we starting this with a deprecated message - shouldn't we just directly use the new non-deprecated flow?

File: src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts [hunk 3]
```diff
@@ -82,6 +82,12 @@ import { isWorkspaceFolder, TaskQuickPickEntry, QUICKOPEN_DETAIL_CONFIG, TaskQui
 import { ILogService } from 'vs/platform/log/common/log';
 import { once } from 'vs/base/common/functional';
 
+// {{ SQL CARBON EDIT }}
+// integration with tasks view panel
+import { ITaskService as ISqlTaskService, TaskStatusChangeArgs } from 'sql/workbench/services/tasks/common/tasksService';
+import { TaskStatus } from 'sql/workbench/api/common/extHostBackgroundTaskManagement';
+import { TaskInfo } from 'azdata';
```
Comment (udeeshagautam) [inline src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts:89]: is it okay to ref a azdata item from VS tree - just checkin if it would cause any code layering warnings to show up?

File: src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts [hunk 4]
```diff
@@ -258,7 +267,8 @@ export abstract class AbstractTaskService extends Disposable implements ITaskSer
 		@ITextModelService private readonly textModelResolverService: ITextModelService,
 		@IPreferencesService private readonly preferencesService: IPreferencesService,
 		@IViewDescriptorService private readonly viewDescriptorService: IViewDescriptorService,
-		@ILogService private readonly logService: ILogService
+		@ILogService private readonly logService: ILogService,
+		@ISqlTaskService private readonly sqlTaskService: ISqlTaskService
```
Comment (udeeshagautam) [inline src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts:271]: so now we can have one task tie in its log/state in activities pane as well as output pane - or it will have to pick one or another?

**Comment by udeeshagautam** [comment]
Looks good @abist - just few questions from my side.

File: src/vs/workbench/api/common/extHost.api.impl.ts (hunk 2 above)
Comment (Charles-Gagnon) [reply src/vs/workbench/api/common/extHost.api.impl.ts:759]: This is from VS Code

File: src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts [hunk 5]
```diff
@@ -2555,9 +2595,16 @@ export abstract class AbstractTaskService extends Disposable implements ITaskSer
 
 		ProblemMatcherRegistry.onReady().then(() => {
 			return this.editorService.saveAll({ reason: SaveReason.AUTO }).then(() => { // make sure all dirty editors are saved
+				// {{SQL CARBON EDIT}}
+				let lastRunTasksViewTask = this.lastRunTasksViewTask;
```
Comment (Charles-Gagnon) [inline src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts:2599]: Why is this necessary? Shouldn't the rerun function below be the one doing this kind of thing?

File: src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts (hunk 4 above)
Comment (Charles-Gagnon) [reply src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts:271]: Yeah - how does this fit into the VS Code tasks system? Are we planning on making the existing ADS tasks more like VS Code tasks in how they're defined and ran?

**Review by Charles-Gagnon — CHANGES_REQUESTED** [review]
Shouldn't this be only enabled for insiders to start with? Or do we think this is going to be complete enough for the November release?

File: src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts (hunk 4 above)
Comment (abist) [reply src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts:271]: The way ADS task works right now will work the way it does. What this will do is introduce the VS Code Task system functionality, and those tasks will also be shown in the ADS Tasks Panel. I didn't want to change the ADS Task system since there are extensions that depend on it.

File: src/vs/workbench/services/configuration/common/configuration.ts [hunk 6]
```diff
@@ -26,7 +28,7 @@ export const TASKS_CONFIGURATION_KEY = 'tasks';
 export const LAUNCH_CONFIGURATION_KEY = 'launch';
 
 export const WORKSPACE_STANDALONE_CONFIGURATIONS = Object.create(null);
-WORKSPACE_STANDALONE_CONFIGURATIONS[TASKS_CONFIGURATION_KEY] = `${FOLDER_CONFIG_FOLDER_NAME}/${TASKS_CONFIGURATION_KEY}.json`;
+WORKSPACE_STANDALONE_CONFIGURATIONS[TASKS_CONFIGURATION_KEY] = `${TASKS_FOLDER_CONFIG_FOLDER_NAME}/${TASKS_CONFIGURATION_KEY}.json`;
```
[the remark this answers is not in the document]
Comment (abist) [reply src/vs/workbench/services/configuration/common/configuration.ts:31]: Yes, this was a feedback from when I presented earlier and had a `.azuredatastudio` folder for the tasks file.  This doesn't deal with ADS Tasks at all

File: src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts (hunk 3 above)
Comment (alanrenmsft) [reply src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts:89]: this happens to be an interface only, not recommended but works since at runtime it will be ignored. azdata module doesn't exist here at runtime.

File: src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts (hunk 5 above)
Comment (abist) [reply src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts:2599]: The ADS Tasks system is different from the VS Code Task system, so to include it in the Tasks View, I had to create another variable.

**Comment by abist** [comment]
@Charles-Gagnon I think the plan for it is to go in November, since we already delayed it from October.

File: src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts [hunk 7]
```diff
@@ -227,6 +233,9 @@ export abstract class AbstractTaskService extends Disposable implements ITaskSer
 	private _waitForSupportedExecutions: Promise<void>;
 	private _onDidRegisterSupportedExecutions: Emitter<void> = new Emitter();
 
+	// {{SQL CARBON EDIT}}
+	private lastRunTasksViewTask: TaskInfo;
```
Comment (alanrenmsft) [inline src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts:237]: this name is kind of confusinng, do you mean recentTask? :)

Comment (abist) [reply src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts:237]: It's actually not the last or recent task, but the last one to show up on the tasks view, so I wanted to make that clear

File: src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts [hunk 8]
```diff
@@ -1473,8 +1482,25 @@ export abstract class AbstractTaskService extends Disposable implements ITaskSer
 
 		const execTask = async (task: Task, resolver: ITaskResolver): Promise<ITaskSummary> => {
 			return ProblemMatcherRegistry.onReady().then(() => {
+				// {{ SQL CARBON EDIT }}
+				const taskNodeId = UUID.generateUuid();
+				let taskInfo: TaskInfo = {
```
Comment (alanrenmsft) [inline src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts:1487]: the task info definition is very specific to database, can we make it more generic?

Comment (abist) [reply src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts:1487]: This is how SQL Tasks are defined. If we want to generalize that, we would need to change the whole SQL Tasks API. This is just showing SQL Tasks under the generic Tasks View, so database is kind of expected for a SQL Task

File: src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts [hunk 9]
```diff
@@ -2555,9 +2595,16 @@ export abstract class AbstractTaskService extends Disposable implements ITaskSer
 
 		ProblemMatcherRegistry.onReady().then(() => {
 			return this.editorService.saveAll({ reason: SaveReason.AUTO }).then(() => { // make sure all dirty editors are saved
+				// {{SQL CARBON EDIT}}
+				let lastRunTasksViewTask = this.lastRunTasksViewTask;
+				lastRunTasksViewTask.taskId = UUID.generateUuid();
+				this.sqlTaskService.createNewTask(lastRunTasksViewTask);
+				this.lastRunTasksViewTask = lastRunTasksViewTask;
```
Comment (alanrenmsft) [inline src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts:2602]: what does this.sqlTaskService.createNewTask do? according to its name it should return an object, right?
also, you assigned this.lastRunTasksViewTask to a local variable and then assign it back to this.lastRunTasksViewTask, is it really necessary? is createNewTask actually updating the lastRunTasksViewTask variable?

File: src/vs/workbench/services/configuration/common/configuration.ts (hunk 1 above)
Comment (alanrenmsft) [reply src/vs/workbench/services/configuration/common/configuration.ts:10]: is this a requirement? it is confusing that azuredatastudio is actually relying on a .vscode folder.

File: src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts (hunk 9 above)
Comment (abist) [reply src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts:2602]: Replied in other comment

**Comment by abist** [comment]
Bump

File: src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts [hunk 10]
```diff
@@ -1546,6 +1573,19 @@ export abstract class AbstractTaskService extends Disposable implements ITaskSer
 				throw new TaskError(Severity.Warning, nls.localize('TaskSystem.active', 'There is already a task running. Terminate it first before executing another task.'), TaskErrors.RunningTask);
```
Comment (Charles-Gagnon) [inline src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts:1573]: Should you be updating the sql task when this happens?

**Review by Charles-Gagnon — APPROVED** [review]

File: src/vs/workbench/services/configuration/common/configuration.ts (hunk 1 above)
Comment (alanrenmsft) [reply src/vs/workbench/services/configuration/common/configuration.ts:10]: @abist any comment here?

**Review by alanrenmsft — APPROVED** [review]
overall looks good to me

File: src/vs/workbench/services/configuration/common/configuration.ts (hunk 1 above)
Comment (abist) [reply src/vs/workbench/services/configuration/common/configuration.ts:10]: this was a follow up from the last time I presented the Task system for demo. I initially had it as .azuredatastudio
